File: Object-C/BlackFireIM/Example/Profile/Controller/BFProfileService.m
//
//  BFProfileService.m
//  BlackFireIM
//
//  Created by benny wang on 2021/4/21.
//

#import "BFProfileService.h"
#import "NSString+Encry.h"
#import <MSIMSDK/MSIMSDK.h>
#import <AFNetworking.h>
#import "MSIMSDK-UIKit.h"

@interface BFProfileService()


@end
@implementation BFProfileService


+ (void)requestIMToken:(NSString *)uid
               success:(void(^)(NSDictionary *dic))succ
                  fail:(void(^)(NSError *error))fail
{
    if (uid == nil) {
        fail(nil);
        return;
    }
    AFHTTPSessionManager *manager = [self ms_manager];
    
    NSString *postUrl = [NSString stringWithFormat:@"%@/user/iminit",[self postUrl]];
    NSMutableDictionary *params = [NSMutableDictionary dictionary];
    [params setValue:uid forKey:@"uid"];
    [params setValue:@(0) forKey:@"ctype"];
    [manager POST:postUrl parameters:params headers:[self ms_header] progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:responseObject options:0 error:nil];
        NSNumber *code = dic[@"code"];
        if (code.integerValue == 0) {
            if (succ) succ(dic[@"data"]);
        }else {
            NSError *err = [NSError errorWithDomain:dic[@"msg"] code:code.integerValue userInfo:nil];
            if (fail) fail(err);
            MSLog(@"%@",err);
        }

        } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {

            if (fail) fail(error);
            MSLog(@"%@",error);
    }];
}

///模拟用户注册
+ (void)userSignUp:(NSString *)phone
          nickName:(NSString *)nickName
            avatar:(NSString *)avatar
              succ:(void(^)(void))succ
            failed:(void(^)(NSError *error))fail
{
    AFHTTPSessionManager *manager = [self ms_manager];
    NSString *postUrl = [NSString stringWithFormat:@"%@/user/reg",[self postUrl]];
    NSMutableDictionary *params = [NSMutableDictionary dictionary];
    [params setValue:phone forKey:@"uid"];
    [params setValue:nickName forKey:@"nick_name"];
    [params setValue:avatar forKey:@"avatar"];
    [params setValue:@(1) forKey:@"gender"];
    [params setValue:@(YES) forKey:@"gold"];
    [params setValue:@(YES) forKey:@"approved"];
    [params setValue:@(NO) forKey:@"disabled"];
    [params setValue:@(NO) forKey:@"blocked"];
    [params setValue:@(NO) forKey:@"hold"];
    [params setValue:@(NO) forKey:@"deleted"];
    [params setValue:@(YES) forKey:@"verified"];
    [manager POST:postUrl parameters:params headers:[self ms_header] progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:responseObject options:0 error:nil];
        NSNumber *code = dic[@"code"];
        if (code.integerValue == 0) {
            if (succ) succ();
        }else {
            NSError *err = [NSError errorWithDomain:dic[@"msg"] code:code.integerValue userInfo:nil];
            if (fail) fail(err);
            MSLog(@"%@",err);
        }
        
        } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
            
            if (fail) fail(error);
            MSLog(@"%@",error);
    }];
}

///修改个人资料
+ (void)requestToEditProfile:(MSProfileInfo *)info
                     success:(void(^)(NSDictionary *dic))succ
                        fail:(void(^)(NSError *error))fail
{
    if (info == nil) {
        fail(nil);
        return;
    }
    AFHTTPSessionManager *manager = [self ms_manager];
    
    NSString *postUrl = [NSString stringWithFormat:@"%@/user/update",[self postUrl]];
    NSMutableDictionary *params = [NSMutableDictionary dictionary];
    [params setValue:info.user_id forKey:@"uid"];
    [params setValue:info.nick_name forKey:@"nick_name"];
    [params setValue:info.avatar forKey:@"avatar"];
    [params setValue:@(info.gender) forKey:@"gender"];
    [params setValue:@(info.gold) forKey:@"gold"];
    [params setValue:@(info.verified) forKey:@"verified"];
    [manager POST:postUrl parameters:params headers:[self ms_header] progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:responseObject options:0 error:nil];
        if ([dic[@"code"]integerValue] == 0) {
            if (succ) succ(dic);
        }else {
            if (fail) fail([[NSError alloc]initWithDomain:dic[@"msg"] code:[dic[@"code"]integerValue] userInfo:nil]);
        }
        } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
            
            if (fail) fail(error);
            MSLog(@"%@",error);
    }];
}

+ (NSDictionary *)ms_header
{
    NSString *secret = @"asfasdasd123";
    NSString *radom = [NSString stringWithFormat:@"%u",arc4random_uniform(1000000)];
    NSString *time = [NSString stringWithFormat:@"%zd",[MSIMTools sharedInstance].adjustLocalTimeInterval/1000/1000];
    NSString *sign = [[NSString stringWithFormat:@"%@%@%@",secret,radom,time] bf_sh1];
    return @{@"nonce":radom,@"timestamp":time,@"sig":sign,@"appid":@"1"};
}

+ (AFHTTPSessionManager *)ms_manager
{
    AFHTTPSessionManager *manager = [AFHTTPSessionManager manager];
    manager.securityPolicy = [AFSecurityPolicy policyWithPinningMode:AFSSLPinningModeNone];
    manager.securityPolicy.allowInvalidCertificates = YES;
    manager.securityPolicy.validatesDomainName = NO;
    manager.requestSerializer = [AFJSONRequestSerializer serializer];
    manager.responseSerializer = [AFHTTPResponseSerializer serializer];
    manager.requestSerializer.timeoutInterval = 5;
    return manager;
}

+ (NSString *)postUrl
{
    BOOL serverType = [[NSUserDefaults standardUserDefaults]boolForKey:@"ms_Test"];
    NSString *host = serverType ? @"https://im.ekfree.com:18788" : @"https://msim1.ekfree.com:18788";
    return host;
}

//@"https://im.ekfree.com:18788"
//@"https://192.168.123.225:18788"
@end

File: Swift/MSIM-Swift/Login/BFProfileService.swift
//
//  BFProfileService.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/7/28.
//

import Foundation
import Moya
import Alamofire
import HandyJSON
import MSIMSDK


public enum ProfileService {
    
    case getIMToken(uid: String)
    case register(phone: String,nickName: String,avatar: String)
    case profileEdit(info: MSProfileInfo)
    case download(url: String,destination: String)
}

extension ProfileService: TargetType {
    public var baseURL: URL {
        switch self {
        case .download(let url,_):
            return URL(string: url)!
        default:
            let serverType = UserDefaults.standard.bool(forKey: "ms_Test")
            let host = serverType ? URL(string: "https://im.ekfree.com:18789")! : URL(string: "https://msim.ekfree.com:18789")!
            return host
        }
    }
    
    public var path: String {
        switch self {
        case .getIMToken:
            return "/user/iminit"
        case .register:
            return "/user/reg"
        case .profileEdit:
            return "/user/update"
        default:
            return ""
        }
    }
    
    public var method: Moya.Method {
        
        switch self {
        case .download:
            return .get
        default:
            return .post
        }
    }
    
    public var sampleData: Data {
        return "{}".data(using: .utf8)!
    }

    public var task: Task {
        
        var params: [String: Any] = [:]
        switch self {
        case .getIMToken(let uid):
            params = ["uid": uid,"ctype": 0] as [String: Any]
            return .requestParameters(parameters: params, encoding: JSONEncoding.default)
        case .register(let phone, let nickName, let avatar):
            params = ["uid": phone,"nick_name": nickName,"avatar": avatar,"gender": 1] as [String: Any]
            return .requestParameters(parameters: params, encoding: JSONEncoding.default)
        case .profileEdit(let info):
            params = ["uid": info.user_id,"nick_name": info.nick_name,"avatar": info.avatar,"gender": info.gender] as [String: Any]
            return .requestParameters(parameters: params, encoding: JSONEncoding.default)
        case .download(_,let savePath):
            let destination: DownloadRequest.Destination = {_,_ in
                return (URL(fileURLWithPath: savePath),[.removePreviousFile,.createIntermediateDirectories])
            }
            return .downloadDestination(destination)
        }
    }
    
    public var headers: [String : String]? {
        switch self {
        case .download:
            return nil
        default:
            let secret = "asfasdasd123";
            let radom = String(format: "%u", arc4random_uniform(1000000));
            let time = String(format: "%zd", MSIMTools.sharedInstance().adjustLocalTimeInterval/1000/1000);
            let sign = String(format: "%@%@%@",secret,radom,time).sha1;
            return ["nonce":radom,"timestamp":time,"sig":sign!,"app_id":"2"]
        }
    }
}


//struct Certificates {
//    static let certificate: SecCertificate = Certificates.certificate(filename: "certificateFileName")
//
//    private static func certificate(filename: String) -> SecCertificate {
//        let filePath = Bundle.main.path(forResource: filename, ofType: "cer") ?? ""
//        let data = try! Data(contentsOf: URL(fileURLWithPath: filePath))
//        let certificate = SecCertificateCreateWithData(nil, data as CFData)!
//        return certificate
//  }
//}

let manager = ServerTrustManager(allHostsMustBeEvaluated: false,evaluators: [:])
let session = Session(serverTrustManager: manager)

let provider = MoyaProvider<ProfileService>(session: session,plugins: [NetworkLoggerPlugin(configuration: NetworkLoggerPlugin.Configuration(logOptions: .verbose))])

extension ProfileService {
    
    
    static func iMTokenAPI(uid: String, success:@escaping (_ result: Any?) -> Void, fail:@escaping (_ error: NSError?) -> Void) {
        
        provider.request(ProfileService.getIMToken(uid: uid)) { result in
            
            switch result {
                case let .success(response):
                    let jsonDic = try! response.mapJSON() as! NSDictionary
                    guard let code = jsonDic["code"] as? Int,let msg = jsonDic["msg"] as? String else {
                        fail(nil)
                        return
                    }
                    if code == 0 {
                        success(jsonDic["data"])
                    }else {
                        let err = NSError(domain: msg, code: code, userInfo: nil)
                        fail(err)
                    }
                case let .failure(error):
                    let err = NSError(domain: error.errorDescription ?? "", code: error.errorCode, userInfo: error.errorUserInfo)
                    fail(err)
            }
        }
    }
    
    static func userRegistAPI(phone: String,nickName: String,avatar: String, success:@escaping (_ result: Any?) -> Void, fail:@escaping (_ error: NSError?) -> Void) {
        
        provider.request(ProfileService.register(phone: phone, nickName: nickName, avatar: avatar)) { result in
            switch result {
                case let .success(response):
                    let jsonDic = try! response.mapJSON() as! NSDictionary
                    guard let code = jsonDic["code"] as? Int,let msg = jsonDic["msg"] as? String else {
                        fail(nil)
                        return
                    }
                    if code == 0 {
                        success(jsonDic["data"])
                    }else {
                        let err = NSError(domain: msg, code: code, userInfo: nil)
                        fail(err)
                    }
                case let .failure(error):
                    let err = NSError(domain: error.errorDescription ?? "", code: error.errorCode, userInfo: error.errorUserInfo)
                    fail(err)
            }
        }
    }
    
    static func userEditAPI(info: MSProfileInfo, success:@escaping (_ result: Any?) -> Void, fail:@escaping (_ error: NSError?) -> Void) {
        
        provider.request(ProfileService.profileEdit(info: info)) { result in
            switch result {
                case let .success(response):
                    let jsonDic = try! response.mapJSON() as! NSDictionary
                    guard let code = jsonDic["code"] as? Int,let msg = jsonDic["msg"] as? String else {
                        fail(nil)
                        return
                    }
                    if code == 0 {
                        success(jsonDic["data"])
                    }else {
                        let err = NSError(domain: msg, code: code, userInfo: nil)
                        fail(err)
                    }
                case let .failure(error):
                    let err = NSError(domain: error.errorDescription ?? "", code: error.errorCode, userInfo: error.errorUserInfo)
                    fail(err)
            }
        }
    }
    
    static func download(sourceUrl: String,savePath: String,progress: ProgressBlock?,success:@escaping (_ result: Any?) -> Void, fail:@escaping (_ error: NSError?) -> Void) {
        
        provider.request(ProfileService.download(url: sourceUrl, destination: savePath),progress: progress) { result in
            
            switch result {
            case .success(_):
                    success(nil)
            case let .failure(error):
                    let err = NSError(domain: error.errorDescription ?? "", code: error.errorCode, userInfo: error.errorUserInfo)
                    fail(err)
            }
        }
    }
}
